ProjFS.Mac/PrjFSKext/PrjFSKext/PerformanceTracing.hpp

```diff
@@ -0,0 +1,171 @@
+#pragma once
+
+#include "PrjFSCommon.h"
+
+#include <mach/mach_time.h>
+#include <IOKit/IOReturn.h>
+
+void PerfTracing_Init();
+void PerfTracing_ProbeInit(PerfTracingProbe* probe);
+void PerfTracing_RecordSample(PerfTracingProbe* probe, uint64_t startTime, uint64_t endTime);
+
+struct IOExternalMethodArguments;
+IOReturn PerfTracing_ExportDataUserClient(IOExternalMethodArguments* arguments);
+
+extern PerfTracingProbe profile_probes[Probe_Count];
+// A scope-based manual instrumentation profiling mechanism.
+// In the simplest case, the time between construction and destruction of the ProfileSample
+// is measured and registered with the probe specified during construction:
+//
+// void MyFunction()
+// {
+//    ProfileSample functionSample(Probe_MyFunction);
+//     // ... The code for which we're measuring the runtime ...
+// } // <-- functionSample goes out of scope here, so that's when timing ends
+//
+//
+// To allow runtimes different code paths in the same scope to be recorded separately, the
+// probe identity can be modified using SetProbe():
+//
+// void MyFunction()
+// {
+//     ProfileSample functionSample(Probe_MyFunction);
+//     // ... The code for which we're measuring the runtime ...
+//     if (specialCase)
+//     {
+//         // We want to be able to distinguish between the runtimes of MyFunction for this
+//         // special case vs "normal" runs.
+//         functionSample.SetProbe(Probe_MyFunctionSpecialCase);
+//         // ... do something potentially expensive here ...
+//     }
+//     // ... more code ...
+// } // <-- Runtime to here will be recorded either under Probe_MyFunction or Probe_MyFunctionSpecialCase
```

Comment:
```Runtime to here will be recorded either under Probe_MyFunction or Probe_MyFunctionSpecialCase```

To confirm, if `specialCase` is true, the runtime will *always* be recorded under `Probe_MyFunctionSpecialCase`?

Reply:
yes